bundles/io/org.eclipse.smarthome.io.http.auth/src/main/java/org/eclipse/smarthome/io/http/auth/internal/AuthenticationHandler.java

```diff
@@ -0,0 +1,160 @@
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Request handler which allows to verify authentication.
+ *
+ * @author Łukasz Dywicki - initial contribution.
+ */
+@Component(configurationPid = "org.eclipse.smarthome.io.http.auth")
+public class AuthenticationHandler implements Handler {
+
+    private static final String AUTHENTICATION_ENABLED = "authentication.enabled";
+    private static final String AUTHENTICATION_ENDPOINT = "authentication.loginUri";
+
+    private final Logger logger = LoggerFactory.getLogger(AuthenticationManager.class);
+
+    private final List<CredentialsExtractor<HttpServletRequest>> extractors = new CopyOnWriteArrayList<>();
+
+    private AuthenticationManager authenticationManager;
+
+    // configuration properties
+    private boolean enabled = true;
+    private String loginUri = "/login";
+
+    @Override
+    public int getPriority() {
+        return HandlerPriorities.AUTHENTICATION;
+    }
+
+    @Override
+    public void handle(final HttpServletRequest request, final HttpServletResponse response,
+            final HandlerContext context) {
+        String requestUri = request.getRequestURI();
+        if (this.enabled && isSecured(requestUri, request.getMethod())) {
+            if (authenticationManager == null) {
+                throw new AuthenticationException("Failed to authenticate request.");
+            }
+
+            int found = 0, failed = 0;
+            for (CredentialsExtractor<HttpServletRequest> extractor : extractors) {
+                Optional<Credentials> extracted = extractor.retrieveCredentials(request);
+                if (extracted.isPresent()) {
+                    found++;
+                    Credentials credentials = extracted.get();
+                    try {
+                        Authentication authentication = authenticationManager.authenticate(credentials);
+                        request.setAttribute(Authentication.class.getName(), authentication);
+                        context.execute(request, response);
+                        return;
+                    } catch (AuthenticationException e) {
+                        failed++;
+                        logger.info("Failed to authenticate using credentials {}", credentials, e);
+                    }
+                }
+            }
+
+            throw new AuthenticationException("Could not authenticate request. Found " + found
+                    + " credentials in request out of which " + failed + " were invalid");
+        }
+    }
+
+    @Override
+    public void handleError(HttpServletRequest request, HttpServletResponse response, HandlerContext context) {
+        Object error = request.getAttribute(HandlerContext.ERROR_ATTRIBUTE);
+
+        if (error instanceof AuthenticationException) {
+            // force client redirect
+            response.setHeader("Location", loginUri);
```

Comment:
Shouldn't you pass the called url as a parameter, so that the form can actually redirect back once the user authenticated successfully?

Reply:
Correct, will update code to support such scenario.